src/cpp/threadpool.cpp

```diff
@@ -0,0 +1,164 @@
+#include <iostream>
+#include <string>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <chrono>
+
+#include "threadpool.h"
+#include "srexception.h"
+
+using namespace SmartRedis;
+using namespace std::chrono_literals;
+
+// TIME_THREADPOOL
+// Enable this flag to display timings for the threadpool activity.
+// Currently, this will be to screen, but eventually it will go to
+// the SmartRedis log
+#undef TIME_THREADPOOL
+
+// C-tor
+ThreadPool::ThreadPool(unsigned int num_threads)
+{
+    // Flag that we're initializing
+    initialization_complete = false;
+
+    // By default, we'll make one thread for each hardware context
+    if (num_threads == 0)
+        num_threads = std::thread::hardware_concurrency();
+
+    // Create worker threads
+	if (num_threads < 1) num_threads = 1; // Force a minimum of 1 thread
+    for (unsigned int i = 0; i < num_threads; i++) {
+        #ifdef TIME_THREADPOOL
+        std::cout << "Kicking off thread " + std::to_string(i) << std::endl;
+        #endif
+        threads.push_back(std::thread(&ThreadPool::perform_jobs, this, i));
+    }
+
+    // Announce that we're open for business
+    shutting_down = false;
+    shutdown_complete = false;
+    initialization_complete = true;
+}
+
+// D-tor
+ThreadPool::~ThreadPool()
+{
+    // Shut down the threadpool if it hasn't already been stopped
+    if (!shutdown_complete) {
+        shutdown();
+    }
+}
+
+// Shut down the thread pool, blocking any further jobs from submission
+void ThreadPool::shutdown()
+{
+    // Make sure initialization is complete before we shut down
+    while (!initialization_complete)
```

Comment:
Here also, should there be a small wait to avoid hammering the loop?

Reply:
If we're shutting down, then there's nothing else we care about doing, so the spin-wait is appropriate